### PR TITLE
Add table aws_lambda_function_event_source_mapping Closes #1872

### LINKED ISSUE #1872: Add table aws_lambda_event_source_mapping
Proposing a new table `aws_lambda_event_source_mapping` which supports the Event Source Mapping Configurations which associate event sources to each Lambda Function. Preparing a PR for this.

**References**

- [EventSourceMappingConfiguration](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/lambda@v1.39.2/types#EventSourceMappingConfiguration)
- [Client.GetEventSourceMapping](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/lambda#Client.GetEventSourceMapping)
- [GetEventSourceMappingInput](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/lambda#GetEventSourceMappingInput)
- [GetEventSourceMappingOutput](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/lambda#GetEventSourceMappingOutput)

Work in progress:

steampipe - per - aws_lambda_event_source_mapping

### PR DESCRIPTION
**Pull request by nickman**
# Integration test logs
<details>
  <summary>Logs</summary>

```
Add passing integration test logs here
```
</details>

# Example query results
<details>
  <summary>Results</summary>

```sql
select
  event_source_arn,
  function_arn,
  function_name,
  last_processing_result,
  parallelization_factor,
  state,
  destination_config
from
  aws_lambda_function_event_source_mapping;
```

|event_source_arn|function_arn|function_name|last_processing_result|parallelization_factor|state|destination_config|
|----------------|------------|-------------|----------------------|----------------------|-----|------------------|
|arn:aws:kinesis:us-east-1:XXXXXXXXXXXXXX:stream/w8imy|arn:aws:lambda:us-east-1:XXXXXXXXXXXXXX:function:eventaudit|eventaudit|OK|9|Enabled|{"OnFailure": {"Destination": "arn:aws:sqs:us-east-1:XXXXXXXXXXXXXX:FailedAuditEvents"}, "OnSuccess": null}|
|arn:aws:kinesis:us-east-1:XXXXXXXXXXXXXX:stream/instruments|arn:aws:lambda:us-east-1:XXXXXXXXXXXXXX:function:eventaudit|eventaudit|OK|10|Enabled|{"OnFailure": {"Destination": null}, "OnSuccess": null}|
|arn:aws:kinesis:us-east-1:XXXXXXXXXXXXXX:stream/partnerProfiles|arn:aws:lambda:us-east-1:XXXXXXXXXXXXXX:function:eventaudit|eventaudit|OK|10|Enabled|{"OnFailure": {"Destination": null}, "OnSuccess": null}|

```sql
select
  runtime,
  count(*)
from
    aws_lambda_function_event_source_mapping, 
    aws_lambda_function
where
    arn = function_arn
group by
  runtime;  
```

|runtime|count|
|-------|-----|
|go1.x|75|
|java11|3|
|java8|264|
|nodejs14.x|1|
|nodejs8.10|1|

```sql
select
    function_name, 
    state, 
    last_modified, 
    state_transition_reason
from
    aws_lambda_function_event_source_mapping
where
    state = 'Disabled';
```

|function_name|state|last_modified|state_transition_reason|
|-------------|-----|-------------|-----------------------|
|NodeKafkaListener|Disabled|2023-08-10 15:35:51.117 -0400|USER_INITIATED|
|fanout|Disabled|2021-02-10 07:27:00.000 -0500|Unrecoverable error|
|eventhandler|Disabled|2022-07-06 12:58:00.000 -0400|Unrecoverable error|
|eventreplicator|Disabled|2022-07-06 12:58:00.000 -0400|Unrecoverable error|
|DynamoHandler|Disabled|2021-06-18 11:33:00.000 -0400|Unrecoverable error|
|fanout|Disabled|2021-02-11 10:18:00.000 -0500|Unrecoverable error|

```sql
select
	jsonb_array_elements_text(
		jsonb_extract_path(self_managed_event_source, 'Endpoints', 'KAFKA_BOOTSTRAP_SERVERS')
	) bootstrap_servers
from
   	aws_lambda_function_event_source_mapping
where 
	function_name = 'NodeKafkaListener'
```

|bootstrap_servers|
|-----------------|
|46.200.141.173:9092|
|46.200.141.174:9092|
|46.200.141.175:9092|
|46.200.141.176:9092|
|46.200.141.177:9092|

</details>

### CODE REVIEW

File: aws/table_aws_lambda_function_event_source_mapping.go [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+	return nil, nil
+}
+
+func getFunctionNameFromArn(ctx context.Context, td *transform.TransformData) (interface{}, error) {
+	arn := *td.HydrateItem.(types.EventSourceMappingConfiguration).FunctionArn
+	parts := strings.Split(arn, ":")
+	return parts[len(parts)-1], nil
+}
+
+// Commented these, and the using columns because the implementation is brittle or non-useful in some cases.
```
Comment (cbruno10) [inline aws/table_aws_lambda_function_event_source_mapping.go:317]: If we don't use these functions, can we remove them and add examples to extract the info, e.g., using the Postgres `split_part()` function?

File: docs/tables/aws_lambda_function_event_source_mapping.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  function_name,
+  last_processing_result,
+  parallelization_factor,
+  state,
+  destination_config
+from
+  aws_lambda_function_event_source_mapping;
+```
+
+### List of lambda functions which disabled
```
Comment (cbruno10) [inline docs/tables/aws_lambda_function_event_source_mapping.md:23]: Is this showing functions that are disabled or disabled event source mappings?

File: docs/tables/aws_lambda_function_event_source_mapping.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    state, 
+    last_modified, 
+    state_transition_reason
+from
+    aws_lambda_function_event_source_mapping
+where
+    state = 'Disabled';
+```
+
+### Count of lambda function event source mappings by runtime engines of the function
```
Comment (cbruno10) [inline docs/tables/aws_lambda_function_event_source_mapping.md:37]: This example seems more relevant to the `aws_lambda_function` table, are there any other examples we can add specific to event source mappings? For instance, showing the data in `source_access_configurations`? Or showing  event sources with a particular filter criteria (from `filter_criteria`)?

File: docs/tables/aws_lambda_function_event_source_mapping.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+select
+  runtime,
+  count(*)
+from
+    aws_lambda_function_event_source_mapping, 
+    aws_lambda_function
+where
+    arn = function_arn
+group by
+  runtime;
```
Comment (cbruno10) [inline docs/tables/aws_lambda_function_event_source_mapping.md:49]:
```suggestion
select
  runtime,
  count(*)
from
  aws_lambda_function_event_source_mapping, 
  aws_lambda_function
where
  arn = function_arn
group by
  runtime;
```

Please format examples according to https://steampipe.io/docs/develop/table-docs-standards#example-query-guidelines

File: docs/tables/aws_lambda_function_event_source_mapping.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+```
+
+### List Kafka Bootstrap Servers from a self-managed Kafka cluster feeding events to a Lambda function
```
Comment (cbruno10) [inline docs/tables/aws_lambda_function_event_source_mapping.md:52]: What does the output of this query look like?

File: docs/tables/aws_lambda_function_event_source_mapping.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
+# Table: aws_lambda_function_event_source_mapping
+
+AWS Lambda is a compute service that lets you run code without provisioning or managing servers.
```
Comment (cbruno10) [inline docs/tables/aws_lambda_function_event_source_mapping.md:3]:
```suggestion
```

I think it's ok if we just describe event source mappings in this table doc (else we'd be describing each service in each table doc)

File: docs/tables/aws_lambda_function_event_source_mapping.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+Event Source Mappings are triggers that route events from services like Kafka, Kinesis, SQS or SNS to the configured AWS Lambda.
+
+## Examples
```
Comment (cbruno10) [inline docs/tables/aws_lambda_function_event_source_mapping.md:6]: Are there any other examples we can add that highlight which fields are relevant to specific resource types? For instance, looking at https://docs.aws.amazon.com/lambda/latest/dg/API_GetEventSourceMapping.html#API_GetEventSourceMapping_ResponseElements, a lot are only relevant to DynamoDB streams, so if there was an example only listing event source mappings using DDB streams and had DDB relevant columns in its `select` statement, that would be easy for users to run and know which columns are relevant.

**Review by cbruno10 — CHANGES_REQUESTED** [review]
@nickman Thanks for opening this PR!

I've left some review comments and questions, can you please have a look and let us know if you have any questions?

File: aws/plugin.go [hunk 8]
```diff
@@ -308,6 +308,7 @@ func Plugin(ctx context.Context) *plugin.Plugin {
 			"aws_lambda_function_metric_errors_daily":                      tableAwsLambdaFunctionMetricErrorsDaily(ctx),
 			"aws_lambda_function_metric_invocations_daily":                 tableAwsLambdaFunctionMetricInvocationsDaily(ctx),
 			"aws_lambda_function":                                          tableAwsLambdaFunction(ctx),
+			EV_SOURCE_MAPPING_TABLE_NAME:                                   tableAwsLambdaFunctionEventSourceMapping(ctx),
```
[the remark this answers is not in the document]
Comment (nickman) [reply aws/plugin.go:311]: Hi @cbruno10
Thanks for the review.
Just to clarify... 
Should the base name be _just_ `aws_lambda_event_source_mapping`
or should I ditch the `lambda` part too... making it `aws_event_source_mapping` ?

[the remark this answers is not in the document]
Comment (cbruno10) [reply aws/plugin.go:311]: @nickman `aws_lambda_event_source_mapping` would be good. We generally use the naming scheme `aws_<service>_<resource>` for table names (this is to help users find resources quickly and eliminate the possibility of conflicts, e.g., if two services have resources with the same name).

**Pull request merged**